delivery_postlogistics/models/stock_picking.py

```diff
@@ -0,0 +1,287 @@
+        Packages are mandatory in this case
+        """
+        for pick in self:
+            pick._set_a_default_package()
+            shipping_labels = pick._generate_postlogistics_label()
+            for label in shipping_labels:
+                pick.attach_shipping_label(label)
+        return True
+
+    def postlogistics_cod_amount(self):
+        """ Return the Postlogistic Cash on Delivery amount of a picking
+
+        If the picking delivers the whole sales order, we use the total
+        amount of the sales order.
+
+        Otherwise, we don't know the value of each picking so we raise
+        an error.  The user has to create packages with the cash on
+        delivery price on each package.
+        """
+        self.ensure_one()
+        order = self.sale_id
+        if not order:
+            return 0.0
+        if len(order) > 1:
+            raise exceptions.Warning(
+                _(
+                    "The cash on delivery amount must be manually specified "
+                    "on the packages when a package contains products "
+                    "from different sales orders."
+                )
+            )
+        order_moves = order.mapped("order_line.procurement_ids.move_ids")
+        picking_moves = self.move_lines
+        # check if the package delivers the whole sales order
+        if order_moves != picking_moves:
+            raise exceptions.Warning(
+                _(
+                    "The cash on delivery amount must be manually specified "
+                    "on the packages when a sales order is delivered "
+                    "in several delivery orders."
+                )
+            )
+        return order.amount_total
+
+    def _generate_postlogistics_label(self, webservice_class=None, package_ids=None):
+        """ Generate labels and write tracking numbers received """
+        self.ensure_one()
+        user = self.env.user
+        company = user.company_id
+        if webservice_class is None:
+            webservice_class = PostlogisticsWebService
+
+        if package_ids is None:
+            packages = self._get_packages_from_picking()
+            packages = sorted(packages, key=attrgetter("name"))
+        else:
+            # restrict on the provided packages
+            package_obj = self.env["stock.quant.package"]
+            packages = package_obj.browse(package_ids)
+
+        web_service = webservice_class(company)
+        res = web_service.generate_label(self, packages, user_lang=user.lang)
+
+        if "errors" in res:
+            raise exceptions.Warning("\n".join(res["errors"]))
+
+        def info_from_label(label):
+            tracking_number = label["tracking_number"]
+            return {
+                "file": base64.b64decode(label["binary"]),
+                "file_type": label["file_type"],
+                "name": tracking_number + "." + label["file_type"],
+            }
+
+        labels = []
```

Comment:
Can you extract the writing of the tracking numbers in another method?

Reply:
done